plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -2723,13 +2891,26 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
             return Optional.empty();
         }
 
+        replaceTableHandle(tableHandle, newHandle);
         return Optional.of(new ConstraintApplicationResult<>(
                 newHandle,
                 newUnenforcedConstraint.transformKeys(ColumnHandle.class::cast),
                 extractionResult.remainingExpression(),
                 false));
     }
 
+    private synchronized void replaceTableHandle(DeltaLakeTableHandle tableHandle, DeltaLakeTableHandle newHandle)
```

Comment:
Can we leave a code comment why replacing the handle is required? Probably, future code readers can't understand this method's motivation.